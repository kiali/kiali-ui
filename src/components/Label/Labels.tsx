import * as React from 'react';
import Label from './Label';
import { style } from 'typestyle';

const SHOW_MORE_TRESHOLD = 2;

interface Props {
  labels?: { [key: string]: string };
}

interface State {
  expanded: boolean;
}

const linkStyle = style({
  float: 'left',
  margin: '7px 2px 2px 3px',
  fontSize: '0.8rem'
});

class Labels extends React.Component<Props, State> {
  constructor(props: Props, state: State) {
    super(props, state);
    this.state = {
      expanded: false
    };
  }

  labelKeys() {
    return Object.keys(this.props.labels || {});
  }

  hasLabels() {
    return this.labelKeys().length > 0;
  }

  hasManyLabels() {
    return this.labelKeys().length > SHOW_MORE_TRESHOLD;
  }

  showItem(i: number) {
    return this.state.expanded || !this.hasManyLabels() || i < SHOW_MORE_TRESHOLD;
  }

  expandLabels = () => {
    this.setState({ expanded: true });
  };

  renderMoreLabelsLink() {
    if (this.hasManyLabels() && !this.state.expanded) {
      return (
        <a key="label_more" className={linkStyle} onClick={this.expandLabels}>
          {' '}
          More labels...
        </a>
      );
    }

    return null;
  }

  renderLabels() {
    return this.labelKeys().map((key, i) => {
      const hideClass = this.showItem(i) ? '' : 'hide';
      return (
        <div key={'label_' + i} className={hideClass}>
          <Label key={'label_' + i} name={key} value={this.props.labels ? this.props.labels[key] : ''} />
        </div>
      );
    });
  }

  renderEmptyLabels() {
    return <span> No labels </span>;
  }

  render() {
    if (this.hasLabels()) {
      return [this.renderLabels(), this.renderMoreLabelsLink()];
    } else {
      return this.renderEmptyLabels();
    }
  }
}

export default Labels;
