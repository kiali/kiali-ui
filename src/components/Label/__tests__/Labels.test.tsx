import * as React from 'react';
import { shallow } from 'enzyme';
import Labels from '../Labels';
import { shallowToJson } from 'enzyme-to-json';

const mockBadge = (labels: { [key: string]: string }) => {
  const component = <Labels labels={labels} />;
  return shallow(component);
};

describe('#Labels render correctly with data', () => {
  it('should render badges with More labels link', () => {
    const wrapper = mockBadge({
      app: 'bookinfo',
      version: 'v1',
      env: 'prod',
      team: 'A'
    });

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render badges without More labels link', () => {
    const wrapper = mockBadge({
      app: 'bookinfo',
      version: 'v1'
    });

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
