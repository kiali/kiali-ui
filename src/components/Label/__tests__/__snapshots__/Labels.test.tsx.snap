// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Labels render correctly with data should render badges with More labels link 1`] = `
Array [
  <div
    className=""
    key="label_0"
  >
    <Label
      key="label_0"
      name="app"
      value="bookinfo"
    />
  </div>,
  <div
    className=""
    key="label_1"
  >
    <Label
      key="label_1"
      name="version"
      value="v1"
    />
  </div>,
  <div
    className=""
    key="label_2"
  >
    <Label
      key="label_2"
      name="env"
      value="prod"
    />
  </div>,
  <div
    className="hide"
    key="label_3"
  >
    <Label
      key="label_3"
      name="team"
      value="A"
    />
  </div>,
  <a
    className="fqiflux"
    onClick={[Function]}
  >
     
    More labels...
  </a>,
]
`;

exports[`#Labels render correctly with data should render badges without More labels link 1`] = `
Array [
  <div
    className=""
    key="label_0"
  >
    <Label
      key="label_0"
      name="app"
      value="bookinfo"
    />
  </div>,
  <div
    className=""
    key="label_1"
  >
    <Label
      key="label_1"
      name="version"
      value="v1"
    />
  </div>,
  "",
]
`;
